Refactor caret animation out of FrameSelection

rr-codes opened the pull request:
#### 6063669dc45d7f10481798a4266f613e4b3cfb5e
<pre>
Refactor caret animation out of FrameSelection
<a href="https://bugs.webkit.org/show_bug.cgi?id=248112">https://bugs.webkit.org/show_bug.cgi?id=248112</a>
rdar://102535758

Reviewed by Simon Fraser and Darin Adler.

Simplifies `FrameSelection` by factoring out the animation logic into its own
class, to fit with how we handle other animations. Instead of `FrameSelection`
owning its own caret blinking timer, the caret is now subject to the repaint cycle
of `Page::updateRendering`.

This PR introduces a new `CaretAnimation` virtual class, with a `CaretAnimationSimple`
subclass to model how the current caret blinking works. This inheritence
hierarchy is modelled after how `ScrollAnimation` works.

The CaretAnimation now owns a timer, which when fired, schedules a rendering
update for the caret to update its appearance. When the rendering update happens,
`CaretAnimation::serviceAnimation` is triggered, which performs the actual
adjustment to the caret&apos;s presentation&apos;s properties. `FrameSelection::invalidateCaretRect`
is then triggered which causes the actual painting of the care to reflect its
new appearance.

Note that because the animation is now tied to `Page::updateRendering`,
`CaretAnimation::serviceAnimation` may be called at any given time, and so the
animation must track when the last time it was called was and account for this.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serviceCaretAnimation):
* Source/WebCore/dom/Document.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::willBeRemovedFromFrame):
(WebCore::FrameSelection::paintCaret):
(WebCore::FrameSelection::setCaretBlinkingSuspended):
(WebCore::FrameSelection::isCaretBlinkingSuspended const):
(WebCore::FrameSelection::caretAnimationDidUpdate):
(WebCore::FrameSelection::document):
(WebCore::FrameSelection::updateAppearance):
(WebCore::FrameSelection::setCaretVisibility):
(WebCore::FrameSelection::setCaretColor):
(WebCore::FrameSelection::caretBlinkTimerFired): Deleted.
(WebCore::FrameSelection::setCaretBlinks): Deleted.
* Source/WebCore/editing/FrameSelection.h:
(WebCore::FrameSelection::currentCaretAnimator):
(WebCore::FrameSelection::currentCaretAnimator const):
(WebCore::FrameSelection::setCaretBlinkingSuspended): Deleted.
(WebCore::FrameSelection::isCaretBlinkingSuspended const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::Page::finalizeRenderingUpdate):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/CaretAnimator.cpp: Added.
(WebCore::CaretAnimator::page const):
(WebCore::CaretAnimator::serviceCaretAnimation):
(WebCore::CaretAnimator::scheduleAnimation):
* Source/WebCore/platform/CaretAnimator.h: Added.
(WebCore::CaretAnimationClient::caretAnimationDidUpdate):
(WebCore::CaretAnimationClient::caretAnimationWillStart):
(WebCore::CaretAnimationClient::caretAnimationDidEnd):
(WebCore::CaretAnimator::CaretAnimator):
(WebCore::CaretAnimator::stop):
(WebCore::CaretAnimator::isActive const):
(WebCore::CaretAnimator::setCaretBlinkingSuspended):
(WebCore::CaretAnimator::isCaretBlinkingSuspended const):
(WebCore::CaretAnimator::didStart):
(WebCore::CaretAnimator::didEnd):
(WebCore::CaretAnimator::timeSinceStart const):
* Source/WebCore/platform/SimpleCaretAnimator.cpp: Added.
(WebCore::SimpleCaretAnimator::SimpleCaretAnimator):
(WebCore::SimpleCaretAnimator::serviceAnimation):
(WebCore::SimpleCaretAnimator::start):
(WebCore::SimpleCaretAnimator::debugDescription const):
* Source/WebCore/platform/SimpleCaretAnimator.h: Added.

Canonical link: <a href="https://commits.webkit.org/257219@main">https://commits.webkit.org/257219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6e202ecdfa471b6db61b37270f0acb2e870f580

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107555 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167819 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102034 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7797 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36071 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90656 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104163 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5863 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84690 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32968 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87732 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89447 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1281 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20879 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1244 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22382 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4974 "Failed to update pull request") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6106 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44833 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2546 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41789 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,119 @@
+
+    virtual String debugDescription() const = 0;
+
+    void fired() {
```
Nit: { should be on new line.

rr-codes replied:
@rniwa This PR isn't meant for review, its just a draft PR for my own sake :)

rniwa commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,119 @@
+    virtual void caretAnimationDidEnd(CaretAnimation&) { }
+};
+
+class CaretAnimation {
```
We usually call these objects CaretAnimationController since it represents a thing that controls animation, not animation itself.

smfr replied:
Agree, or `CaretAnimator`. This isn't identical to `ScrollAnimation` because this drives its own updates.

rniwa commented on Source/WebCore/platform/CaretAnimationSimple.cpp:
```diff
@@ -0,0 +1,53 @@
+
+namespace WebCore {
+
+CaretAnimationSimple::CaretAnimationSimple(CaretAnimationClient& client)
```
We should call this SimpleCaretAnimationController instead.

rr-codes replied:
I'm following the existing convention that ScrollAnimation uses

rniwa commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,119 @@
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    enum class Type {
+        Simple
```
Why is this Type thing needed if we're going to use an inheritance?

rr-codes replied:
@rniwa This is modeled after `ScrollAnimation`, which also has this `Type` enum. It is needed so that the `is` function can work properly with the polymorphism.

smfr replied:
You can do that with virtual functions too.

rniwa replied:
Please use a virtual function to do this.

rr-codes replied:
how would this be done with virtual functions? Should I have a virtual function that returns a `Type` and use that instead of dynamicDowncast / is ?

rniwa replied:
For starters, it's weird that FrameSelection has to downcast just to call `caretPaint`. A better design would have been to have a virtual caretPaint on the interface object / class itself.

rniwa replied:
The way virtual function based type check works is that there will be a virtual function that returns Type or alternatively a bunch of isX where X is the name/type of a subclass.

darinadler replied:
Or we figure out why the caller needs to check the "type" and make a virtual function that answers the real question it is asking.

rniwa commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
Since we're copying the code from FrameSelection.cpp, we need to inherit the copyright information from that file instead.

rniwa commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1895,6 +1904,34 @@ void CaretBase::paintCaret(const Node& node, GraphicsContext& context, const Lay
+void FrameSelection::setCaretBlinkingSuspended(bool suspended)
+{
+    if (m_currentCaretAnimation)
+        m_currentCaretAnimation->setCaretBlinkingSuspended(suspended);
```
This seems broken. We can't forget the fact caret blinking had been suspended
just because the animation wasn't scheduled at that point.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -75,9 +75,14 @@
 #include "TypingCommand.h"
 #include "VisibleUnits.h"
+#include "platform/CaretAnimation.h"
+#include "wtf/Assertions.h"
```
These need `<>` instead of `“”`.

rr-codes replied:
@darinadler This is just a draft PR for my own sake, so it has a bunch of style errors and stuff and it doesn't need to be reviewed until I make it a non-draft PR

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -75,9 +75,14 @@
 #include <stdio.h>
 #include <wtf/text/CString.h>
 #include <wtf/text/TextStream.h>
+#include <CaretAnimationSimple.h>
```
This should be `”“` instead of `<>`.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -75,9 +75,14 @@
 #include "StyleTreeResolver.h"
 #include "TypingCommand.h"
 #include "VisibleUnits.h"
+#include "platform/CaretAnimation.h"
```
Is this the correct include style for things from Platform?

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -163,22 +168,19 @@ static inline bool isPageActive(Document* document)
-#if ENABLE(TEXT_CARET)
-    , m_caretBlinkTimer(*this, &FrameSelection::caretBlinkTimerFired)
-#endif
+// #if ENABLE(TEXT_CARET)
```
Should delete.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -195,6 +197,9 @@ FrameSelection::FrameSelection(Document* document)
 #endif
 }
 
+FrameSelection::~FrameSelection() {
```
Should use `= default`.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1895,6 +1904,34 @@ void CaretBase::paintCaret(const Node& node, GraphicsContext& context, const Lay
+
+bool FrameSelection::isCaretBlinkingSuspended() const
+{
+    if (m_currentCaretAnimation)
```
Use `&&` instead?

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1895,6 +1904,34 @@ void CaretBase::paintCaret(const Node& node, GraphicsContext& context, const Lay
+    return false;
+}
+
+void FrameSelection::caretAnimationDidUpdate(CaretAnimation &)
```
Should not have space before `&`.

rr-codes commented on Source/WebCore/platform/CaretAnimationSimple.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    const auto caretBlinkInterval = RenderTheme::singleton().caretBlinkInterval();
+
+    if (currentTime - m_lastTimeCaretPaintWasToggled >= caretBlinkInterval - (caretBlinkInterval * 0.01)) {
```
this seems weird, but the delta time is always close to `caretBlinkInterval` but never exactly `caretBlinkInterval`. I'm not sure if there's a better way of handling this.

Another reason this seems bad to me is that it introduces `m_lastTimeCaretPaintWasToggled` even though CaretAnimation already has a `m_lastAnimationFrameTimestamp`

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6904,6 +6904,11 @@ void Document::serviceRequestAnimationFrameCallbacks()
 
+void Document::serviceCaretAnimation()
+{
+    m_selection->currentCaretAnimation()->serviceCaretAnimation(domWindow()->frozenNowTimestamp());
```
Can we be sure that all these pointers are non-null?

rr-codes replied:
Yes; `m_selection` is a `UniqueRef`, and currentCaretAnimation is always initialized in the FrameSelection constructor

pxlcoder replied:
You could use `selection()` instead of `m_selection` to make it clear it's a reference.

I don't think `domWindow()` is guaranteed to be non-null though.

smfr commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -268,6 +272,8 @@ class FrameSelection : private CaretBase {
     void disassociateLiveRange();
     void updateFromAssociatedLiveRange();
 
+    std::unique_ptr<CaretAnimation>& currentCaretAnimation() { return m_currentCaretAnimation; }
```
Is it ever non-null? Can you return a reference?

rniwa replied:
This should return CaretAnimation* or CaretAnimation&, not unique_ptr.

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -163,24 +166,20 @@ static inline bool isPageActive(Document* document)
-    , m_caretBlinks(true)
 #endif
 {
+    m_currentCaretAnimation = makeUnique<CaretAnimationSimple>(*this, m_document.get());
```
You could do this in the initializer list above (as long as it comes after m_document initialization.

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1641,7 +1642,8 @@ void FrameSelection::willBeRemovedFromFrame()
 
 #if ENABLE(TEXT_CARET)
-    m_caretBlinkTimer.stop();
+    if (m_currentCaretAnimation)
```
When is `m_currentCaretAnimation` null?

rr-codes replied:
never; I'll remove this check

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2229,16 +2250,15 @@ void FrameSelection::updateAppearance()
 
-        if (!m_caretPaint) {
-            m_caretPaint = true;
+        if (!m_currentCaretAnimation->visible()) {
```
Simpler to just call `m_currentCaretAnimation->setVisible(true)` unconditionally

rr-codes replied:
@smfr does this mean `invalidateCaretRect` should also be called unconditionally? Right now, this if statement is used so that the caret is only invalidated if its not already in the desired state

smfr replied:
I mean check the visibility state inside the call to setVisible() and early return.

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2299,31 +2319,18 @@ void FrameSelection::setCaretVisibility(CaretVisibility visibility, ShouldUpdate
 #if ENABLE(TEXT_CARET)
-    if (m_caretPaint) {
-        m_caretPaint = false;
+    if (m_currentCaretAnimation->visible()) {
```
Simpler to just call `m_currentCaretAnimation->setVisible(false)` unconditionally

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1839,7 +1841,11 @@ void CaretBase::invalidateCaretRect(Node* node, bool caretRectChanged)
 void FrameSelection::paintCaret(GraphicsContext& context, const LayoutPoint& paintOffset, const LayoutRect& clipRect)
 {
-    if (m_selection.isCaret() && m_caretPaint && m_selection.start().deprecatedNode())
+    bool caretPaint = true;
```
`caretPaint` is a weird name and I don't know what it means. Does it mean "shouldPaintCaret"?

smfr commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,129 @@
+    Page* page() const;
+
+    const Type m_animationType;
+    bool m_isActive { false };
```
Re-order these members to optimize packing (put the bools next to each other for example).

smfr commented on Source/WebCore/platform/CaretAnimation.h:
```diff
@@ -0,0 +1,129 @@
+    Timer m_caretBlinkTimer;
+    CaretAnimationClient& m_client;
+    bool m_isCaretBlinkingSuspended = false;
+    Document* m_document;
```
It would be better if this didn't store a raw document pointer; go via the client.

rniwa replied:
Please use WeakPtr or go through client. r- because of this.

rr-codes replied:
does this mean that the CaretAnimationClient should have a `virtual Document* document() = 0` function?

rniwa replied:
That or this member variable becomes `WeakPtr`.

smfr commented on Source/WebCore/platform/CaretAnimationSimple.h:
```diff
@@ -0,0 +1,56 @@
+    void setVisible(bool visible) override { m_caretPaint = visible; }
+
+private:
+    bool m_caretPaint = true;
```
This needs a new name; it reflects the on/off state of the blink, so maybe an enum would be clearer.

rr-codes replied:
how's
```
    enum class BlinkState : bool {
        Off, On
    }
 ```
?

smfr commented on Source/WebCore/platform/CaretAnimationSimple.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace WebCore {
+
+class CaretAnimationSimple final: public CaretAnimation {
```
`SimpleCaretAnimator` perhaps.

smfr commented on Source/WebCore/platform/CaretAnimationSimple.h:
```diff
@@ -0,0 +1,56 @@
+    CaretAnimationSimple(CaretAnimationClient&, Document*);
+    virtual ~CaretAnimationSimple();
+
+    void serviceAnimation(ReducedResolutionSeconds) override;
```
All the `override` methods can be private.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1904,6 +1908,8 @@ void Page::finalizeRenderingUpdate(OptionSet<FinalizeRenderingUpdateFlags> flags
     }
 #endif
 
+    m_renderingUpdateRemainingSteps.last().remove(RenderingUpdateStep::CaretAnimation);
```
Why do we need this here?

rr-codes replied:
because every RenderingUpdateStep must be removed at some point (Page has an assertion that checks this)

smfr replied:
But `runProcessingStep`does the `remove` for you. Not sure why you need it here. If you're asserting, something else is happening.

rniwa commented on Source/WebCore/platform/CaretAnimation.cpp:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
Again, copyright should match that of FrameSelection.cpp since we're copying code here.

rniwa reviewed (changes requested):
r-

darinadler commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -111,7 +112,7 @@ class DragCaretController : private CaretBase {
 };
 
-class FrameSelection : private CaretBase {
+class FrameSelection : private CaretBase, public WebCore::CaretAnimationClient {
```
This does not need a `WebCore::` prefix. It’s inside the `WebCore` namespace.

darinadler commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -198,9 +201,13 @@ class FrameSelection : private CaretBase {
 
+    void caretAnimationDidUpdate(CaretAnimator&) final;
+
+    Document* document() final;
```
Please make these overrides `private`.

darinadler commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -268,6 +274,10 @@ class FrameSelection : private CaretBase {
 
+    CaretAnimator& currentCaretAnimator() { return *m_currentCaretAnimator; }
+
+    const CaretAnimator& currentCaretAnimator() const { return *m_currentCaretAnimator; }
```
Not sure why we need the word "current" in this name. Are there other caret animators?

darinadler commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -335,21 +343,17 @@ class FrameSelection : private CaretBase {
     SelectionRevealMode m_selectionRevealMode { SelectionRevealMode::DoNotReveal };
     AXTextStateChangeIntent m_selectionRevealIntent;
 
+    std::unique_ptr<CaretAnimator> m_currentCaretAnimator;
```
This should be `UniqueRef`. Also should remove the word "current" from this name.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+#include "Document.h"
+#include "ReducedResolutionSeconds.h"
+#include "Timer.h"
+#include <wtf/FastMalloc.h>
```
Since `Timer.h` will pull this in, this include is not needed.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+
+#include "Document.h"
```
We should be able to just forward-declare `Document`. I don’t think we need to include the whole header.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+namespace WebCore {
+
+class Page;
+class CaretAnimator;
```
Alphabetical order, please.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+        , m_caretBlinkTimer(*this, &CaretAnimator::scheduleAnimation)
+    { }
+
+    virtual ~CaretAnimator() = default;
```
I wish I understood why we need a base class and `SimpleCaretAnimator`. What is the other derived class? Coming in the future perhaps? I don’t think this is a good design pattern, even though we apparently have something else with this structure.

rr-codes replied:
@darinadler Please see the associated Radar (rdar://102535758) for more details on the rationale. @smfr suggested using this design pattern for this purpose

darinadler replied:
OK, it's for future expansion. Understood.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace WebCore {
+
+class SimpleCaretAnimator final: public CaretAnimator {
```
Space after `final` before colon.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,56 @@
+class SimpleCaretAnimator final: public CaretAnimator {
+public:
+    SimpleCaretAnimator(CaretAnimationClient&);
+    virtual ~SimpleCaretAnimator();
```
Should just leave this out. It will get generated.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,56 @@
+
+    bool visible() const override { return m_blinkState == BlinkState::On; }
+
+    void setVisible(bool visible) override { m_blinkState = visible ? BlinkState::On : BlinkState::Off; }
```
We should mark these `final` instead of `override` for WebKit coding style. Also, I would leave out all these blank lines.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,56 @@
+
+    void setVisible(bool visible) override { m_blinkState = visible ? BlinkState::On : BlinkState::Off; }
+
+    BlinkState m_blinkState = BlinkState::On;
```
I know others suggested this design, but I think bool `m_isVisible` would be better, myself.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+
+    virtual void start(ReducedResolutionSeconds currentTime) = 0;
+
+    virtual void stop()
```
Should not be virtual.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+        didEnd();
+    }
+
+    virtual bool isActive() const { return m_isActive; }
```
Should not be virtual.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+    Page* page() const;
+
+    CaretAnimationClient& m_client;
+    Timer m_caretBlinkTimer;
```
Since this class is about carets, we don’t need the word "caret" in the names. This can be `m_blinkTimer`.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+    ReducedResolutionSeconds m_startTime;
+    ReducedResolutionSeconds m_lastAnimationFrameTimestamp;
+    bool m_isActive { false };
+    bool m_isCaretBlinkingSuspended { false };
```
`m_isBlinkingSuspended`

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+    virtual String debugDescription() const = 0;
+
+    void setCaretBlinkingSuspended(bool suspended) { m_isCaretBlinkingSuspended = suspended; }
+    bool isCaretBlinkingSuspended() const { return m_isCaretBlinkingSuspended; }
```
This is the caret animator, so the word "caret" is not needed. Should be `setBlinkingSuspended` and `isBlinkingSuspended`.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+
+    virtual void serviceAnimation(ReducedResolutionSeconds) = 0;
+
+    void serviceCaretAnimation(ReducedResolutionSeconds);
```
Naming here is unclear. How is `serviceAnimation` in a caret animator different from `serviceCaretAnimation`? Need names that make the distinction clear.

Also, since the virtual function is only to be called from the other function, it should be `private`.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+
+    virtual void caretAnimationDidUpdate(CaretAnimator&) { }
+    virtual void caretAnimationWillStart(CaretAnimator&) { }
+    virtual void caretAnimationDidEnd(CaretAnimator&) { }
```
We should omit these. No client is overriding these. We can add them later if we have a client that needs them.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,116 @@
+    void setCaretBlinkingSuspended(bool suspended) { m_isCaretBlinkingSuspended = suspended; }
+    bool isCaretBlinkingSuspended() const { return m_isCaretBlinkingSuspended; }
+
+    virtual bool visible() const = 0;
```
I suggest we name this `isVisible`.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -69,12 +71,14 @@
 #include "StyleTreeResolver.h"
 #include "TypingCommand.h"
 #include "VisibleUnits.h"
+#include "dom/Document.h"
```
This include is incorrect. It should just be `Document.h` but also I am surprised it’s needed.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.cpp:
```diff
@@ -0,0 +1,67 @@
+void SimpleCaretAnimator::serviceAnimation(ReducedResolutionSeconds currentTime)
+{
+    if (isCaretBlinkingSuspended() && visible())
+        return;
```
I don’t understand this check. If caret blinking is suspended but the caret is _not_ visible, then why would we still want to run the code below? Confusing without a comment.

rr-codes replied:
this is existing code; it's so that the caret will always be visible when blinking is suspended. I will add a comment.

darinadler replied:
I remember the original now. The comment will help and I also think we can improve how this works in the future. When blinking is turned off I don't think we should need to service an animation to make the cursor appear.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.cpp:
```diff
@@ -0,0 +1,67 @@
+    if (isCaretBlinkingSuspended() && visible())
+        return;
+
+    const auto caretBlinkInterval = RenderTheme::singleton().caretBlinkInterval();
```
The `const` here is not helpful. It just means we happen to not change this local. I suggest we omit it. Most locals aren’t changed but we typically don’t put `const` on them.

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    const auto caretBlinkInterval = RenderTheme::singleton().caretBlinkInterval();
+
+    if (currentTime - m_lastTimeCaretPaintWasToggled >= caretBlinkInterval - (caretBlinkInterval * 0.01)) {
```
Why the 1% fudge factor here? The old code didn’t need this kind of strangeness. Does this enhance caret blinking in some way?

rr-codes replied:
It's because previously, `m_isVisible` was just toggled every time the timer was fired. However now, it uses the Page::updateRendering / scheduleRenderingUpdate pipeline, which means that the timing will not be exactly the interval of the timer, and the function may be called at any time by external events, so the delta time must now be consulted to determine if it should be toggled or not.

darinadler replied:
Yes but the 1% here makes it clear we have racy code that won't do the right thing in all cases. We need to do this some way that doesn't require a fudge factor.

rr-codes replied:
Ah ok, I'll work on fixing this, thanks!

darinadler replied:
I think we want to subtract by a more predictable quantum based on timer resolution. Not 1% of the interval, which works by accident.

smfr replied:
This might be timer slop due to alignment, or maybe it's the "reduced resolution" part of `ReducedResolutionSeconds`. If you wait at least 500ms do you always get an update after that?

rr-codes replied:
Instead of going through Page::updateRendering, I'll just have the SimpleAnimator blink using the timer directly since then a threshold factor is no longer needed

rr-codes replied:
@darinadler @smfr It actually works now with  Page::updateRendering without the threshold factor, since now the timer is a one-shot timer instead of a repeating one!

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.cpp:
```diff
@@ -0,0 +1,67 @@
+String SimpleCaretAnimator::debugDescription() const
+{
+    TextStream textStream;
+    textStream << "SimpleCaretAnimator " << this << " active " << isActive() << " blink state = " << (m_blinkState == BlinkState::On ? "BlinkState::On" : "BlinkState::Off");
```
I suggest just "on" and "off" for serializing the blink state. But could we also come up with a different way to refer to this than "blink state", which seems kind of oblique?

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2229,16 +2249,15 @@ void FrameSelection::updateAppearance()
-            m_caretPaint = true;
+        if (!caretAnimator().isVisible()) {
+            caretAnimator().setVisible(true);
             invalidateCaretRect();
```
Shouldn't it be the responsibility of the `caretAnimator` to call `caretAnimationDidUpdate` if visibility changes, which then triggers the invalidation? Ideally there'd only be one place that calls `invalidateCaretRect()`.

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2299,31 +2318,18 @@ void FrameSelection::setCaretVisibility(CaretVisibility visibility, ShouldUpdate
-        m_caretPaint = false;
+    if (caretAnimator().isVisible()) {
+        caretAnimator().setVisible(false);
         invalidateCaretRect();
```
Ditto

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,49 @@
+
+class SimpleCaretAnimator final : public CaretAnimator {
+public:
+    SimpleCaretAnimator(CaretAnimationClient&);
```
Should add `explicit`.

smfr commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -111,7 +112,7 @@ class DragCaretController : private CaretBase {
 };
 
-class FrameSelection : private CaretBase {
+class FrameSelection : private CaretBase, public CaretAnimationClient {
```
FrameSelection can probably be `final`

darinadler commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,49 @@
+    bool isVisible() const final { return m_isVisible; }
+    void setVisible(bool visible) final { m_isVisible = visible; }
+
+    bool m_isVisible = true;
```
Should use the `{ }` syntax here like above. Arbitrary but consistent in most of WebKit.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,113 @@
+        m_blinkTimer.stop();
+    }
+
+    Seconds timeSinceStart(ReducedResolutionSeconds currentTime) const
```
This function is not used yet. I suggest removing it unless you need it for future code.

smfr commented on Source/WebCore/platform/SimpleCaretAnimator.h:
```diff
@@ -0,0 +1,49 @@
+    String debugDescription() const final;
+
+    bool isVisible() const final { return m_isVisible; }
+    void setVisible(bool visible) final { m_isVisible = visible; }
```
I'd expect this to trigger an invalidation.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,113 @@
+class CaretAnimator {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    CaretAnimator(CaretAnimationClient& client)
```
Needs `explicit`. Also, this constructor should be `protected` instead of `public`.

darinadler commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,113 @@
+ */
+
+#pragma once
+
```
Extra blank line here.

darinadler reviewed (approved):
Ideas for further tweaks. Nothing important though.

smfr commented on Source/WebCore/platform/CaretAnimator.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void CaretAnimator::serviceCaretAnimation(ReducedResolutionSeconds timestamp)
+{
+    if (timestamp < m_lastAnimationFrameTimestamp) {
```
I think this will cause a serious power regression.

Let's say something (like an an animated GIF) is triggering 100ms rendering updates, and this caret timer is 500ms. `serviceCaretAnimation()` will get called for each rendering update triggered by the GIF (so every 100ms). But here, you're asking for another rendering update right away, so we'll do rendering updates at 16ms intervals!

smfr replied:
I misread this code; seems like `timestamp < m_lastAnimationFrameTimestamp` will never happen. Just delete `m_lastAnimationFrameTimestamp`

smfr commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -0,0 +1,112 @@
+private:
+    void scheduleAnimation();
+
+    bool m_isActive { false };
```
I think this is too easily conflated with "is the caret visible at all" (i.e. in the blinking state). Something like `enum class CaretBlinkState { On, Off }` would be clearer.

rr-codes committed:
Refactor caret animation out of FrameSelection
https://bugs.webkit.org/show_bug.cgi?id=248112
rdar://102535758

Reviewed by Simon Fraser and Darin Adler.

Simplifies `FrameSelection` by factoring out the animation logic into its own
class, to fit with how we handle other animations. Instead of `FrameSelection`
owning its own caret blinking timer, the caret is now subject to the repaint cycle
of `Page::updateRendering`.

This PR introduces a new `CaretAnimation` virtual class, with a `CaretAnimationSimple`
subclass to model how the current caret blinking works. This inheritence
hierarchy is modelled after how `ScrollAnimation` works.

The CaretAnimation now owns a timer, which when fired, schedules a rendering
update for the caret to update its appearance. When the rendering update happens,
`CaretAnimation::serviceAnimation` is triggered, which performs the actual
adjustment to the caret's presentation's properties. `FrameSelection::invalidateCaretRect`
is then triggered which causes the actual painting of the care to reflect its
new appearance.

Note that because the animation is now tied to `Page::updateRendering`,
`CaretAnimation::serviceAnimation` may be called at any given time, and so the
animation must track when the last time it was called was and account for this.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::serviceCaretAnimation):
* Source/WebCore/dom/Document.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::willBeRemovedFromFrame):
(WebCore::FrameSelection::paintCaret):
(WebCore::FrameSelection::setCaretBlinkingSuspended):
(WebCore::FrameSelection::isCaretBlinkingSuspended const):
(WebCore::FrameSelection::caretAnimationDidUpdate):
(WebCore::FrameSelection::document):
(WebCore::FrameSelection::updateAppearance):
(WebCore::FrameSelection::setCaretVisibility):
(WebCore::FrameSelection::setCaretColor):
(WebCore::FrameSelection::caretBlinkTimerFired): Deleted.
(WebCore::FrameSelection::setCaretBlinks): Deleted.
* Source/WebCore/editing/FrameSelection.h:
(WebCore::FrameSelection::currentCaretAnimator):
(WebCore::FrameSelection::currentCaretAnimator const):
(WebCore::FrameSelection::setCaretBlinkingSuspended): Deleted.
(WebCore::FrameSelection::isCaretBlinkingSuspended const): Deleted.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::Page::finalizeRenderingUpdate):
(WebCore::operator<<):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/CaretAnimator.cpp: Added.
(WebCore::CaretAnimator::page const):
(WebCore::CaretAnimator::serviceCaretAnimation):
(WebCore::CaretAnimator::scheduleAnimation):
* Source/WebCore/platform/CaretAnimator.h: Added.
(WebCore::CaretAnimationClient::caretAnimationDidUpdate):
(WebCore::CaretAnimationClient::caretAnimationWillStart):
(WebCore::CaretAnimationClient::caretAnimationDidEnd):
(WebCore::CaretAnimator::CaretAnimator):
(WebCore::CaretAnimator::stop):
(WebCore::CaretAnimator::isActive const):
(WebCore::CaretAnimator::setCaretBlinkingSuspended):
(WebCore::CaretAnimator::isCaretBlinkingSuspended const):
(WebCore::CaretAnimator::didStart):
(WebCore::CaretAnimator::didEnd):
(WebCore::CaretAnimator::timeSinceStart const):
* Source/WebCore/platform/SimpleCaretAnimator.cpp: Added.
(WebCore::SimpleCaretAnimator::SimpleCaretAnimator):
(WebCore::SimpleCaretAnimator::serviceAnimation):
(WebCore::SimpleCaretAnimator::start):
(WebCore::SimpleCaretAnimator::debugDescription const):
* Source/WebCore/platform/SimpleCaretAnimator.h: Added.

Canonical link: https://commits.webkit.org/257219@main
